packages/meditrak-app-server/src/__tests__/__integration__/sync/CountChangesRoute.test.ts

```diff
@@ -0,0 +1,299 @@
+
+  it('throws an error if no auth header provided', async () => {
+    const response = await app.get('changes/count');
+    expect(response.statusCode).toEqual(500);
+    expect(response.body.error).toMatch(/.*Authorization header required/);
+  });
+
+  it('should return a number under the key "changeCount"', async () => {
+    const response = await app.get('changes/count', {
+      headers: {
+        Authorization: authHeader,
+      },
+    });
+    expect(response.statusCode).toEqual(200);
+    expect(typeof response.body.changeCount).toBe('number');
+  });
+
+  it('should return the total number of update changes with no "since"', async () => {
+    const correctChangeCount = await models.meditrakSyncQueue.count({ type: 'update' });
+    const response = await app.get('changes/count', {
+      headers: {
+        Authorization: authHeader,
+      },
+    });
+    expect(response.body).toEqual({ changeCount: correctChangeCount });
+  });
+
+  it('should return the correct number of changes since "since" if updates are made', async () => {
+    const since = Date.now();
+    const numberOfQuestionsToAdd = randomIntBetween(1, 20);
+    await oneSecondSleep();
+    for (let i = 0; i < numberOfQuestionsToAdd; i++) {
+      await upsertQuestion(models);
+    }
+
+    // Wait for the triggers to have properly added the changes to the queue
+    await models.database.waitForAllChangeHandlers();
+    const response = await app.get('changes/count', {
+      headers: {
+        Authorization: authHeader,
+      },
+      query: {
+        since,
+      },
+    });
+    expect(response.body).toEqual({ changeCount: numberOfQuestionsToAdd });
+  });
+
+  it('should return the correct number of changes since "since" if updates and deletes are made', async () => {
+    // Note: sync skips redundant deletes, i.e. any 'delete' records that reflect the deletion of a
+    // record that the client has never seen are not synced to that client
+
+    // Note that throughout this test we sleep before and after taking the timestamps as there is
+    // overlap due to ids not being very fine grained
+
+    // Add some questions
+    const timestampBeforeFirstUpdate = Date.now();
+    await oneSecondSleep();
+    const numberOfQuestionsToAddInFirstUpdate = randomIntBetween(1, 20);
+    const newQuestionsInFirstUpdate = [];
+    for (let i = 0; i < numberOfQuestionsToAddInFirstUpdate; i++) {
+      newQuestionsInFirstUpdate[i] = await upsertQuestion(models);
+    }
+
+    // Add some more questions
+    await oneSecondSleep();
+    const timestampBeforeSecondUpdate = Date.now();
+    await oneSecondSleep();
+    const numberOfQuestionsToAddInSecondUpdate = randomIntBetween(1, 20);
+    const newQuestionsInSecondUpdate = [];
+    for (let i = 0; i < numberOfQuestionsToAddInSecondUpdate; i++) {
+      newQuestionsInSecondUpdate[i] = await upsertQuestion(models);
+    }
+
+    // Delete some of the questions added in the first update
+    await oneSecondSleep();
```

Comment:
These second long waits do add up a tiny bit... tests can take 30-60 seconds on my local machine. Personally I feel it's fine, but in an ideal world it might be worth finding a way to avoid them

Reply:
Yeah...definitely an anti pattern I'm not proud of. We also run the risk of it being flakey based on timing differences - the long wait time mitigates that risk, but obviously comes with its own downside!

Reply:
Yeah, there probably is a way we could work around this if needs be. But the tests still run fast enough imo 👍